ivy/func_wrapper.py

```diff
@@ -293,6 +293,21 @@ def _check_in_nested_sequence(sequence, value=None, _type=None):
             )
 
 
+def _get_preferred_device(args, kwargs):
+    # When new arrays are created, they should be created on the same device as
+    # existing array inputs. If a device is specified as a kwarg, create them there.
+    # If not, scan for any other inputs which are already arrays and use the device
+    # of the first one found (unless we're in soft device mode).
+    device = None
+    if "device" in kwargs and kwargs["device"] is not None:
+        return device
+    if not ivy.soft_device_mode:
+        for arg in args + list(kwargs.values()):
```

Comment:
I think a better way to do this is by using `_get_first_array` which uses `nested_argwhere`. Since the input arrays can be passed in a list.
See implementation of `infer_device` for reference.